### PR TITLE
KAFKA-15970: Copy over KIP-951 tests in FetcherTest.java to FetchRequestManagerTest.java

### PR DESCRIPTION
**Pull request by msn-tldr**
Copying over tests added to FetcherTest.java in [PR](https://github.com/apache/kafka/pull/14685/files#diff-2745b4306d343fd2a5abbf6314c0f8d01d07d7140ab989894ca1a0a17100f02a) to FetchRequestManagerTest.java, so KIP-951 changes have coverage with new consumer being worked as described here 
https://cwiki.apache.org/confluence/display/KAFKA/Consumer+threading+refactor+project+overview

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 1]
```diff
@@ -3126,6 +3127,207 @@ public void testCorruptMessageError() {
...
+        assertTrue(partitionRecords.containsKey(tp0));
+        assertTrue(partitionRecords.containsKey(tp1));
+
+        // Verify that preferred read replica is set for both tp0 & tp1.
+        Node selected = fetcher.selectReadReplica(tp0, Node.noNode(), time.milliseconds());
+        assertEquals(nodeId0.id(), selected.id());
+        selected = fetcher.selectReadReplica(tp1, Node.noNode(), time.milliseconds());
+        assertEquals(nodeId0.id(), selected.id());
+
+        // Next fetch returns an error(due to leadership change) and new leader info is returned. The new leader is a new node, id = 999.
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: This comment seems quite out of place. The vital line of code is 3283.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 2]
```diff
@@ -3126,6 +3127,207 @@ public void testCorruptMessageError() {
...
     }
 
+
+    /**
+     * Test the scenario that FetchResponse returns with an error indicating leadership change for the partition, but it
+     * does not contain new leader info(defined in KIP-951).
+     */
+    @ParameterizedTest
+    @EnumSource(value = Errors.class, names = {"FENCED_LEADER_EPOCH", "NOT_LEADER_OR_FOLLOWER"})
+    public void testWhenFetchResponseReturnsWithALeaderShipChangeErrorButNoNewLeaderInformation(Errors error) {
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: `...ALeadershipChangeError...` please.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java [hunk 3]
```diff
@@ -3126,6 +3127,207 @@ public void testCorruptMessageError() {
...
+                .setLastStableOffset(FetchResponse.INVALID_LAST_STABLE_OFFSET)
+                .setLogStartOffset(0)
+                .setRecords(nextRecords));
+        client.prepareResponseFrom(FetchResponse.of(Errors.NONE, 0, INVALID_SESSION_ID, partitions), nodeId0);
+        networkClientDelegate.poll(time.timer(0));
+        partitionRecords = fetchRecords();
+        assertFalse(partitionRecords.containsKey(tp0));
+        assertTrue(partitionRecords.containsKey(tp1));
+
+        // Validate metadata is unchanged, as response has missing new leader info.
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java]: It's not really missing new leader info, which sounds like a problem. It's just that KIP-951 provides a way to attach the new leader info to the response, and this first test shows the case where that is not being done.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
It's a bit hard to follow, but the overall idea of the tests seems sound.

**Comment by msn-tldr** [comment]
@AndrewJSchofield I have update the comments in the test to make it more readable, hope it helps.

Also i have made same changes to FetcherTest where these tests were originally added.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java [hunk 4]
```diff
@@ -3496,53 +3501,52 @@ public void testWhenFetchResponseReturnsWithALeaderShipChangeErrorButNoNewLeader
...
         selected = fetcher.selectReadReplica(tp1, Node.noNode(), time.milliseconds());
         assertEquals(nodeId0.id(), selected.id());
 
-        // Next fetch returns an error(due to leadership change) but new leader info is missing, for tp0.
-        // For tp1 fetch returns with no error.
+        // Send next fetch request.
         assertEquals(1, sendFetches());
         assertFalse(fetcher.hasCompletedFetches());
-        // Verify that metadata-update isn't requested
+        // Verify that metadata-update isn't requested as metadata is considered upto-date.
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java:3516]: "up to date"

**Review by AndrewJSchofield — APPROVED** [review]
lgtm

**Review by kirktrue — APPROVED** [review]
LGTM, assuming the tests pass 😄

**Review by wcarlson5 — APPROVED** [review]

**Comment by msn-tldr** [comment]
So JDK 17 build failed with following and its unrelated.
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14916/5/pipeline/11/

```
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [a99ad330-91c1-4b36-98c1-035d6d718e13 port:35031, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].

	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)

	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)

	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:103)

	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)

	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)

	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)

Caused by: java.net.ConnectException: Connection refused

	at java.base/sun.nio.ch.Net.pollConnect(Native Method)

	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)

	at java.base/sun.nio.ch.SocketChannelImpl.finishTimedConnect(SocketChannelImpl.java:1141)

	at java.base/sun.nio.ch.SocketChannelImpl.blockingConnect(SocketChannelImpl.java:1183)

	at java.base/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:98)

	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)

	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)

	... 5 more
.....
.....

* What went wrong:

Execution failed for task ':streams:upgrade-system-tests-10:checkstyleTest'.

> A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction

   > Failed to run Gradle Worker Daemon

      > Process 'Gradle Worker Daemon 4' finished with non-zero exit value 1
```

**Comment by wcarlson5** [comment]
I'll rerun it and see if it gets any further :)

**Comment by msn-tldr** [comment]
@wcarlson5 thanks! The failures are unrelated in the new build
https://ge.apache.org/s/ul2ost2737xbg/tests/overview?outcome=FAILED

**Comment by msn-tldr** [comment]
@wcarlson5 thanks!
